src/util/utils.ts

```diff
@@ -181,65 +181,87 @@ export const getUid = (): string => {
 export interface RecordSubscribeArguments { callback: (data: any) => void, path?: string, triggerNow?: boolean }
-  /**
-   * Creates a map based on the types of the provided arguments
-   */
+
+/**
+ * Creates a map based on the types of the provided arguments
+ */
 export const normalizeSetArguments = (args: IArguments, startIndex: number = 0): RecordSetArguments => {
-    let result
+  let result
 
-    if (args.length === startIndex + 1) {
-      result = { data: args[startIndex], path: undefined, callback: undefined }
+  const isRootData = (data: any) => data !== undefined && typeof data === 'object'
+  const isNestedData = (data: any) => data !== undefined && typeof data !== 'function'
+  const isPath = (path: any) => path !== undefined && typeof path === 'string'
+  const isCallback = (callback: any) => typeof callback === 'function'
+
+  if (args.length === startIndex + 1) {
+    result = {
+      path: undefined,
+      data: isRootData(args[startIndex]) ? args[startIndex] : undefined,
+      callback: undefined
     }
+  }
 
-    if (args.length === startIndex + 2) {
-      if (typeof args[startIndex] === 'string' && typeof args[startIndex + 1] === 'object') {
-        result = { path: args[startIndex], data: args[startIndex + 1], callback: undefined }
-      }
-      if (typeof args[startIndex] === 'object' && typeof args[startIndex + 1] === 'function') {
-        result = { path: undefined, data: args[startIndex], callback: args[startIndex + 1] }
-      }
+  if (args.length === startIndex + 2) {
+    result = { path: undefined, data: undefined, callback: undefined }
+    if (!isCallback(args[startIndex + 1]) && isNestedData(args[startIndex + 1])) {
+      result.path = isPath(args[startIndex]) ? args[startIndex] : false
     }
 
-    if (args.length === startIndex + 3) {
-      result = { path: args[startIndex + 0], data: args[startIndex + 1], callback: args[startIndex + 2]}
+    if (isPath(args[startIndex])) {
+      result.data = isNestedData(args[startIndex + 1]) ? args[startIndex + 1] : undefined
+    } else {
+      result.data = isRootData(args[startIndex]) ? args[startIndex] : undefined
     }
 
-    if (result) {
-      if (result.path !== undefined && typeof result.path !== 'string' || result.path.length === 0) {
-        throw Error ('Invalid set path argument')
-      }
-      if (result.callback !== undefined && typeof result.callback !== 'function') {
-        throw Error ('Invalid set callback argument')
-      }
-      if (result.data === undefined) {
-        throw Error ('Invalid set data argument')
-      }
-      return result
+    if (!isPath(args[startIndex])) {
+      result.callback = isCallback(args[startIndex + 1]) ? args[startIndex + 1] : false
     }
+  }
 
-    throw Error ('Invalid set arguments')
+  if (args.length === startIndex + 3) {
+    result = {
+      path: isPath(args[startIndex]) ? args[startIndex] : false,
```

Comment:
How is path false?

Reply:
those `false` values in this case are a way to trigger precise error messages, as `undefined` can be an acceptable case for args.length cases

Reply:
Cool, makes sense now